### PR TITLE
Separate seeds and random number generators

### PR DESCRIPTION
**Pull request by wenzeslaus**
Random number generators for Model can be seeded in Config using one seed, multiple seeds, or one seed but incremented for each of the different generators.

On most places where standard C++ generator was used, custom generator provider class is now used instead and the functions pick what generator they need. Places which are general or reused at different places take the standard C++ generator.

The generator provider classes have also the standard C++ generator interface and can be used wherever a standard C++ generator can be used as long as the configuration allows it.

Generator is no longer part of Simulation, but generator provider is included in Model. Simulation takes it as a parameter of each function which needs a generator similarly to other places in PoPS and C++ library where generators are used.

Config can read text, accept vector of seeds, or std::map can be assigned to its random_seeds attribute.

A lot of tests needed code update because either generator provider object is now needed in the given context (not just seed) or different type is needed (provider, not standard C++ generator).

Text key-value parsing is general. Potentially reusable.

Date now reports the invalid text. (Unrelated but included.)

The new and old core tests and r.pops.spread tests pass. A new test in r.pops.spread tests that enabling and disabling feature does not change results if feature does not have other effects besides consuming random numbers (ncsu-landscape-dynamics/r.pops.spread/pull/61).

### CODE REVIEW

**Comment by wenzeslaus** [comment]
This is ready for review with tests and documentation. Description updated accordingly.

**Review by ChrisJones687 — APPROVED** [review]

File: include/pops/date.hpp
[the remark this answers is not in the document]
Comment (wenzeslaus) [reply include/pops/date.hpp:1]: Motivated by me providing invalid input :-)

File: include/pops/generator_provider.hpp [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+    }
+
+    result_type operator()()
+    {
+        return general_generator_();
+    }
+
+    void discard(unsigned long long n)
+    {
+        general_generator_.disard(n);
```
Comment (petrasovaa) [inline include/pops/generator_provider.hpp:185]: Typo

File: include/pops/generator_provider.hpp [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+ * Depending on how the object is seeded, result is single generator or
+ * multiple independent generators.
+ *
+ * Satisfies UniformRandomBitGenerator, so it can be used in place of
+ * standard generators. This makes testing of other components easier
+ * as this generator object can be used directly or, more importantly,
+ * standard generator can be used in its place.
+ *
+ * However, unlike the simple generator for single seed, this will throw
+ * an exception if used directly as UniformRandomBitGenerato, but the
```
Comment (petrasovaa) [inline include/pops/generator_provider.hpp:359]:
```suggestion
 * an exception if used directly as UniformRandomBitGenerator, but the
```

File: include/pops/generator_provider.hpp [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+ * object was seeded with multiple seeds.
+ */
+template<typename Generator>
+class RandomNumberGeneratorProvider
+{
+public:
+    /**
+     * Seeds first generator with the seed and then each subsequent generator with
+     * seed += 1. *isolated* decides if single generator is created or if multiple
+     * isolated generators are created and seed is incremeneted as needed.
```
Comment (petrasovaa) [inline include/pops/generator_provider.hpp:369]:
```suggestion
     * isolated generators are created and seed is incremented as needed.
```

File: include/pops/generator_provider.hpp [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+/** Default generator provider to be used for tests and development */
+using DefaultSingleGeneratorProvider =
+    SingleGeneratorProvider<std::default_random_engine>;
+
+/** Generator provider providing multiple isolated generators
+ *
+ * All ways of seeding result is multiple independent generators.
+ */
+template<typename Generator>
+class IsolatedRandomNumberGeneratorProvider
```
Comment (petrasovaa) [inline include/pops/generator_provider.hpp:201]: Consider renaming to MultiRandom...

File: include/pops/simulation.hpp [hunk 5]
```diff
@@ -928,7 +943,8 @@ class Simulation
         bool weather,
         DispersalKernel& dispersal_kernel,
         const std::vector<std::vector<int>>& suitable_cells,
-        double establishment_probability = 0.5)
+        double establishment_probability,
```
Comment (petrasovaa) [inline include/pops/simulation.hpp:946]: This is changing somewhat unrelated code, not sure about the implications

**Review by petrasovaa — CHANGES_REQUESTED** [review]
Some typos. I didn't review the tests.

File: include/pops/generator_provider.hpp (hunk 1 above)
Comment (wenzeslaus) [reply include/pops/generator_provider.hpp:185]: Two places fixed.

File: include/pops/generator_provider.hpp (hunk 3 above)
Comment (wenzeslaus) [reply include/pops/generator_provider.hpp:369]: Two places fixed.

File: include/pops/generator_provider.hpp (hunk 4 above)
Comment (wenzeslaus) [reply include/pops/generator_provider.hpp]: Renamed to Multi... together with variables named "isolated". I also made the error messages more consistent with the documentation using "multiple isolated generators".

**Review by petrasovaa — APPROVED** [review]
